src/map.rs

```diff
@@ -423,36 +455,91 @@ impl MapData {
                     ),
             );
 
-        // Add image
-        let (base64_image, viewbox) = match self
+        // Determine background/viewbox
+        let background = self
             .generate_background_image()
-            .map_err(|err| PyValueError::new_err(err.to_string()))?
-        {
-            Some(data) => data,
-            None => return Ok(None),
-        };
-        let image = Image::new()
-            .set("x", viewbox.min_x)
-            .set("y", viewbox.min_y)
-            .set("width", viewbox.width)
-            .set("height", viewbox.height)
-            .set("style", "image-rendering: pixelated")
-            .set("href", format!("data:image/png;base64,{base64_image}"));
-
-        let mut document = Document::new()
-            .set("viewBox", viewbox.to_svg_viewbox())
-            .add(defs)
-            .add(image);
+            .map_err(|err| PyValueError::new_err(err.to_string()))?;
+
+        let mut document = Document::new().add(defs);
+        let (viewbox_opt, bg_img): (Option<ViewBox>, Option<String>) =
+            if let Some((base64_image, viewbox)) = background {
+                document = document.set("viewBox", viewbox.to_svg_viewbox(self.rotation_deg));
+                (Some(viewbox), Some(base64_image))
+            } else if !self.outlines.is_empty()
+                || !self.areas.is_empty()
+                || !self.block_lines.is_empty()
+            {
+                if let Some(vb) =
+                    viewbox_from_shapes(&self.outlines, &self.areas, &self.block_lines)
+                {
+                    document = document.set("viewBox", vb.to_svg_viewbox(self.rotation_deg));
+                    (Some(vb), None)
+                } else {
+                    return Ok(None);
+                }
+            } else {
+                return Ok(None);
+            };
+
+        // Create layers: map_layer (rotated), ui_layer (not rotated)
+        let mut map_layer = Group::new();
+        let mut ui_layer = Group::new();
+        if let Some(vb) = viewbox_opt.as_ref() {
+            if self.rotation_deg != 0.0 {
+                let cx = vb.min_x as f32 + (vb.width as f32 / 2.0);
+                let cy = vb.min_y as f32 + (vb.height as f32 / 2.0);
+                map_layer = map_layer.set(
+                    "transform",
```

Comment:
You are already calculating the viewbox and some parts with rotation in mind. We should calculate everything instead of using transform here, as otherwise the code is strange